presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,575 @@
+        if ((r3 != 0 && (l3 | l2 | l1) != 0)
+                || (r2 != 0 && (l3 | l2) != 0)
+                || (r1 != 0 && l3 != 0)) {
+            throwOverflowException();
+        }
+
+        long product;
+
+        product = (r0 & LONG_MASK) * (l0 & LONG_MASK);
+        int z0 = (int) product;
+
+        product = (r0 & LONG_MASK) * (l1 & LONG_MASK)
+                + (r1 & LONG_MASK) * (l0 & LONG_MASK)
+                + (product >> 32);
+        int z1 = (int) product;
+
+        product = (r0 & LONG_MASK) * (l2 & LONG_MASK)
+                + (r1 & LONG_MASK) * (l1 & LONG_MASK)
+                + (r2 & LONG_MASK) * (l0 & LONG_MASK)
+                + (product >> 32);
+        int z2 = (int) product;
+
+        product = (r0 & LONG_MASK) * (l3 & LONG_MASK)
+                + (r1 & LONG_MASK) * (l2 & LONG_MASK)
+                + (r2 & LONG_MASK) * (l1 & LONG_MASK)
+                + (r3 & LONG_MASK) * (l0 & LONG_MASK)
+                + (product >> 32);
+        int z3 = (int) product;
+        if ((product >>> 32) != 0) {
+            throwOverflowException();
+        }
+
+        boolean negative = isNegative(left) ^ isNegative(right);
+        pack(result, z0, z1, z2, z3, negative);
+
+        throwIfOverflows(z0, z1, z2, z3);
+    }
+
+    public static boolean overflows(Slice value, int precision)
+    {
+        return precision < MAX_PRECISION && compareUnsigned(value, POWERS_OF_TEN[precision]) >= 0;
+    }
+
+    public static int compare(Slice left, Slice right)
+    {
+        boolean leftStrictlyNegative = isStrictlyNegative(left);
+        boolean rightNegative = isNegative(right);
+
+        if (leftStrictlyNegative != rightNegative) {
+            return leftStrictlyNegative ? -1 : 1;
+        }
+        else {
+            return compareUnsigned(left, right) * (leftStrictlyNegative ? -1 : 1);
+        }
+    }
+
+    public static int compare(long leftRawLo, long leftRawHi, long rightRawLo, long rightRawHi)
+    {
+        boolean leftStrictlyNegative = isStrictlyNegative(leftRawLo, leftRawHi);
+        boolean rightNegative = isNegative(rightRawLo, rightRawHi);
+
+        if (leftStrictlyNegative != rightNegative) {
+            return leftStrictlyNegative ? -1 : 1;
+        }
+        else {
+            return compareUnsigned(leftRawLo, leftRawHi, rightRawLo, rightRawHi) * (leftStrictlyNegative ? -1 : 1);
+        }
+    }
+
+    public static int compareUnsigned(Slice left, Slice right)
+    {
+        long leftHi = getLong(left, 1);
+        long rightHi = getLong(right, 1);
+        if (leftHi != rightHi) {
+            return Long.compareUnsigned(leftHi, rightHi);
+        }
+
+        long leftLo = getLong(left, 0);
+        long rightLo = getLong(right, 0);
+        if (leftLo != rightLo) {
+            return Long.compareUnsigned(leftLo, rightLo);
+        }
+
+        return 0;
+    }
+
+    public static int compareUnsigned(long leftRawLo, long leftRawHi, long rightRawLo, long rightRawHi)
```

Comment:
saving one letter on 8 letter variable? Maybe rename this method to:

`public static int cmpUn(long lrl, long lrh, long rrl, long rrh)` ?